### PR TITLE
Add timeout to requests

### LINKED ISSUE #120: requests need timeout
For time-critical applications, it would be great if all of the request calls could use a provided timeout.

### PR DESCRIPTION
**Pull request by AnssiAhola**
Adds (optional) request timeout to the client as requested by @aw-was-here in #120 

Adds properties `connect_timeout` and `read_timeout` to base class `Fetcher`
Will use those properties and pass them to [`requests.request`](https://requests.readthedocs.io/en/latest/api/#requests.request) function as a tuple

These new properties are `None` by default as to not "break" existing apps etc.
Note. They are also None by default if not provided to the `requests.request` function [see link](https://requests.readthedocs.io/en/latest/user/advanced/#timeouts), kind of surprised this hasnt come up sooner :smile:

Usage:
```python
client = Client(my_client_name,user_token=my_token)
client.set_timeout(read=0.5)

user = client.user("MrJuusto")
for folder in user.collection_folders:
    for release in folder.releases:
        print(release)

>>> Traceback (most recent call last):
>>> ...
>>> requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='api.discogs.com', port=443): Read timed out. (read timeout=0.5)
```

Closes #120 

@alifhughes @JOJ0 Feedback is appreciated :smiley: 

@aw-was-here
If you want to test it for yourself before release, that would be appreciated as well :sweat_smile: 
[See our docs](https://python3-discogs-client.readthedocs.io/en/latest/contribution.html#testing-an-unreleased-feature) on how to do that if you're not familiar already

### CODE REVIEW

**Comment by JOJ0** [comment]
That's awesome news @aw-was-here, your testing will help us a lot! BTW I dig your "music info showing to listeners" project, DJ myself, pretty cool stuff you built there!

**Comment by AnssiAhola** [comment]
I was thinking, should we intercept `ReadTimeout` and `ConnectTimeout` exceptions and throw our own?
Or maybe just document this feature so that it's clear where these exceptions come from and when they are thrown?

Speaking of which I should probably write some documentation :sweat_smile:

File: discogs_client/client.py [hunk 1]
```diff
@@ -193,3 +194,20 @@ def backoff_enabled(self, value: bool) -> None:
...
+
+    def set_timeout(self,
+                    connect: Union[int, float] = 5,
+                    read: Union[int, float] = 10) -> None:
+        """Set request timeout parameters
+
+        Args:
+            connect (Union[int, float], optional): Time in seconds after which connect will timeout. Defaults to 5.
+            read (Union[int, float], optional): Time in seconds after which request will timeout. Defaults to 10.
+        """
```
Comment (JOJ0) [inline discogs_client/client.py:211]: This looks like google docstring style, in other places we are using another syntax, I think it's called the numpy style. Have a look at https://github.com/joalla/discogs_client/blob/master/discogs_client/fetchers.py#L23-L50

I'm fine with google style as well, I actually would prefer it but we should keep it consistent throughout the project.

**Comment by JOJ0** [comment]
> I was thinking, should we intercept `ReadTimeout` and `ConnectTimeout` exceptions and throw our own? Or maybe just document this feature so that it's clear where these exceptions come from and when they are thrown?

Having our own exceptions is alwaysnice IMHO but I don't know if it makes any sense. @alifhughes we'd like to know your opinion on that topic :-)

Here's an overview of the exceptions we currently have implemented: https://python3-discogs-client.readthedocs.io/en/latest/discogs_client.html#module-discogs_client.exceptions

> Speaking of which I should probably write some documentation sweat_smile

I'm not sure yet where these docs should have their place. Since it's done at the very start / while intiializing the Client, somewhere here maybe: https://python3-discogs-client.readthedocs.io/en/latest/quickstart.html

but I'm not sure if we should polute the quickstart chapter with this slightly advanced topic. But for now: Why not, put it here and I'll think about it and move it perhaps, before we put out the next release.

**Comment by JOJ0** [comment]
Ah and: I am currently running the discogs_client instance I'm using for my github.com/beetbox/beets library from this branch, so I'll get a little real-life testing going.

File: discogs_client/client.py (hunk 1 above)
Comment (AnssiAhola) [reply discogs_client/client.py:211]: Ahh, good point.
This was autogenerated in this style by some VSCode extension (i guess), I'll convert it to numpy style.

**Comment by AnssiAhola** [comment]
I think this feature falls in line with the rate limiting and it has a separate page [here](https://python3-discogs-client.readthedocs.io/en/latest/requests_rate_limit.html)
But I'll write it in the quickstart first and we'll move it if necessary.

**Comment by aw-was-here** [comment]
> That's awesome news @aw-was-here, your testing will help us a lot! BTW I dig your "whats now playing" project, DJ myself, pretty cool stuff you built there!

Thanks! 

So I was able to do some testing and everything appeared to be ok, minus an issue I don't believe is related (current source tree has a lot of big changes and so hit what I believe is an unrelated edge case).  I'll have another chance to test things out with the latest change tomorrow.

**Comment by aw-was-here** [comment]
FWIW, I've verified the connect timeout works for my use case by pulling the network cable. :smile:   Many thanks for adding this feature!

**Comment by alifhughes** [comment]
Hey guys! Great work @AnssiAhola and thanks for those comments too @JOJ0, this feature looks very handy indeed. For me it looks okay, glad to see tests there. 

As per the throwing our own exceptions, in this case I don't think it is particularly worth it as the exception back is pretty descriptive:
> requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='api.discogs.com', port=443): Read timed out. (read timeout=0.5)

My only comment would be why have `def timeout()` return a tuple? It works, but seems slightly unintuitive as it is returning two separate values. I'd prefer to just have the properties named and it returns the single values, such as:

```python
   @property
    def connection_timeout(self):
        """Return current client connection timeout"""
        return self._fetcher.connect_timeout

    @property
    def read_timeout(self):
        """Return current client read timeout"""
        return self._fetcher.read_timeout
```

I think it is nicer to keep them separate as we're using them separately (not as a tuple) elsewhere in the code. What do you think?

Great work nonetheless guys, glad to see it!

**Comment by alifhughes** [comment]
> > My only comment would be why have `def timeout()` return a tuple?
> 
> Good point and not sure why I ended up on that. I'll seperate them. Thanks for the feedback!

👍 remember to update the tests, once the build is passing I'll approve

**Review by alifhughes — APPROVED** [review]
🚀

**Comment by AnssiAhola** [comment]
Sorry for fumbling with the tests, on my secondary machine where the local env is not setup correctly so it "make a fix" -> "push" ->"hope tests pass" 😅

**Comment by AnssiAhola** [comment]
I'll merge this if there are no objections?
BTW @JOJ0 did you want to move the documentation from quickstart?

**Comment by JOJ0** [comment]
@AnssiAhola hmmm, actually it's not too bad in this place. It's pretty straight forward, can stay there I think. What do you think @alifhughes and @prcutler? Do you see a better place for it?

**Comment by alifhughes** [comment]
I think quickstart is fine imo, only other place would be to make a separate section like the API rate limiting configuration, but I don't mind 👍

File: docs/source/quickstart.md [hunk 2]
```diff
@@ -77,3 +77,18 @@ release = d.release(1)
...
 
+## Request timeouts
+
+By default the {class}`.Client` does not timeout requests.
+
+You can enable request timeouts like so
+
+```python
+timeout_in_seconds = 5
+client.timeout(
```
Comment (th3hamm0r) [inline docs/source/quickstart.md:88]: Thx for this new api!
This should be `set_timeout(...)` and not `timeout(...)`.

Comment (AnssiAhola) [reply docs/source/quickstart.md:88]: Nice catch, thanks! I'll fix it.
